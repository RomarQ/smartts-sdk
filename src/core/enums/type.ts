export enum TypeAtom {
    record = 'record',
    variant = 'variant',
    chain_id = 'chain_id',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    option = 'option',
    map = 'map',
    big_map = 'bigmap',
    signature = 'signature',
    operation = 'operation',
    never = 'never',
    tuple = 'tuple',
    lambda = 'lambda',
    ticket = 'ticket',
    sapling_state = 'sapling_state',
    contract = 'contract',
    sapling_transaction = 'sapling_transaction',
}

export default TypeAtom;
