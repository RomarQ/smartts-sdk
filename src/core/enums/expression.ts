export enum ExpressionAtom {
    // Math
    add = 'add',
    mul = 'mul',
    sub = 'sub',
    truediv = 'truediv',
    // Comparison
    eq = 'eq',
    neq = 'neq',
    lt = 'lt',
    gt = 'gt',
    le = 'le',
    ge = 'ge',
    //
    items = 'items',
    update_map = 'update_map',
    getItem = 'getItem',
    getItemDefault = 'getItemDefault',
    getItemMessage = 'getItemMessage',
    attr = 'attr',
    params = 'params',
    getLocal = 'getLocal',
    iter = 'iter',
    // Blockchain property
    amount = 'amount',
    balance = 'balance',
    chain_id = 'chain_id',
    level = 'level',
    now = 'now',
    self = 'self',
    self_address = 'self_address',
    sender = 'sender',
    source = 'source',
    total_voting_power = 'total_voting_power',
    voting_power = 'voting_power',
    // Operation
    operations = 'operations',
    transfer = 'transfer',
    set_delegate = 'set_delegate',
    create_contract = 'create_contract',
    // List instructions
    cons = 'cons',
    //
    contract = 'contract',
    // Variant
    openVariant = 'openVariant',
    isVariant = 'isVariant',
    cases_arg = 'cases_arg',
    variant_arg = 'variant_arg',
}

export default ExpressionAtom;
