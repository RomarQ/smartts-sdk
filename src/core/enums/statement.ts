export enum StatementAtom {
    verify = 'verify',
    ifBlock = 'ifBlock',
    elseBlock = 'elseBlock',
    forGroup = 'forGroup',
    whileBlock = 'whileBlock',
    set = 'set',
    defineLocal = 'defineLocal',
    result = 'result',
    failwith = 'failwith',
    match = 'match',
    match_cases = 'match_cases',
    set_type = 'set_type',
}

export default StatementAtom;
