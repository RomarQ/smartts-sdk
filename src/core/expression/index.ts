export * from './literal';
export * from './expression';
export * from './logic';
