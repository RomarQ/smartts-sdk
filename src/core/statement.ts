import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import type { LineInfo } from '../misc/utils';

export class Statement implements IStatement {
    args;

    constructor(public name: string, ...args: (IExpression | IType | LineInfo | string | ILiteral<unknown>)[]) {
        this.args = args || [];
    }

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}
