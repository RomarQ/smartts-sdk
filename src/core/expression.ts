import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import { GetProperty, LambdaArgument } from '../expression/variables';
import { capitalizeBoolean, LineInfo, parenthesis } from '../misc/utils';
import ValueAtom from './enums/literal';
import TypeAtom from './enums/type';

export class Expression<T extends ValueAtom> implements IExpression<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;
    private args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return GetProperty(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}

export class LiteralExpression<T extends ValueAtom> implements ILiteral<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;

    constructor(private name: T, private values: (number | string | boolean | IExpression)[], private line: LineInfo) {}

    toString() {
        switch (this.name) {
            case ValueAtom.unit:
                return parenthesis(this.name);
            case ValueAtom.nat:
            case ValueAtom.int:
            case ValueAtom.mutez:
            case ValueAtom.bytes:
            case ValueAtom.bool:
            case ValueAtom.string:
            case ValueAtom.chain_id_cst:
            case ValueAtom.timestamp:
            case ValueAtom.bls12_381_fr:
            case ValueAtom.bls12_381_g1:
            case ValueAtom.bls12_381_g2:
            case ValueAtom.key:
            case ValueAtom.key_hash:
            case ValueAtom.signature:
            case ValueAtom.address:
                return `(literal (${this.name} ${this.values.join(' ')}) ${this.line})`;
            case ValueAtom.list:
            case ValueAtom.set:
            case ValueAtom.tuple:
                return `(${this.name} ${this.line} ${this.values.join(' ')})`;
            case ValueAtom.Some:
            case ValueAtom.None:
                return `(${ValueAtom.variant} "${this.name}" ${this.values.join(' ') || `(${ValueAtom.unit})`} ${
                    this.line
                })`;
            default:
                return `(${this.name} ${this.values.join(' ')} ${this.line})`;
        }
    }
}

export class RecordLiteral implements ILiteral<ValueAtom.record> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as ValueAtom.record;

    constructor(private fields: Record<string, IExpression>, private line: LineInfo) {}

    private buildFields = (fields: Record<string, IExpression>) => {
        return Object.entries(fields).map(([field, expr]) => `(${field} ${expr.toString()})`);
    };

    toString(): string {
        return `(${TypeAtom.record} ${this.line} ${this.buildFields(this.fields).join(' ')})`;
    }
}

export class MapLiteral<T extends ValueAtom.map | ValueAtom.big_map> implements ILiteral<T> {
    _isExpression = true as const;
    _type = {} as T;
    type = {} as IType;

    constructor(
        private prim: T,
        private rows: IExpression[][],
        keyType: IType,
        valueType: IType,
        private line: LineInfo,
    ) {}

    private buildEntry = ([key, value]: IExpression[]) => {
        return `(${key.toString()} ${value.toString()})`;
    };

    toString() {
        return `(${this.prim} ${this.line} ${this.rows.map(this.buildEntry).join(' ')})`;
    }
}

export class LambdaLiteral implements ILiteral<ValueAtom.lambda> {
    _isExpression = true as const;
    // Used for type checking
    _type = ValueAtom.lambda as const;
    type = {} as unknown as IType;
    static idCounter = 0;
    private identifier: number;
    private withStorage?: 'read-write' | 'read-only';
    private withOperations = false;

    constructor(
        private statements: IStatement[],
        private inType: IType,
        private argumentName: string,
        private line: LineInfo,
    ) {
        this.identifier = ++LambdaLiteral.idCounter;
    }

    public get id(): Readonly<number> {
        return this.identifier;
    }

    public setInputType(type: IType) {
        this.inType = type;
        return this;
    }

    public code(buildStatements: (arg: IExpression) => IStatement[]) {
        const param = LambdaArgument(this.argumentName, this.inType, this.id, this.line);
        this.statements = buildStatements(param);
        return this;
    }

    toString() {
        return `(${ValueAtom.lambda} ${this.id} ${this.withStorage || 'None'} ${capitalizeBoolean(
            this.withOperations,
        )} "${this.argumentName}" ${this.line} (${this.statements.join(' ')}))`;
    }
}
