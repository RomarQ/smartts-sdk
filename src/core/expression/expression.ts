import { proxy } from '../../misc/proxy';
import { LineInfo } from '../../misc/utils';
import { IExpression } from '../../typings/expression';
import { ILiteral } from '../../typings/literal';
import { IType } from '../../typings/type';

export class Expression implements IExpression {
    _isExpression = true as const;
    args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string | ILiteral<unknown>)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return Expression.getAttr(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    static getAttr = (from: IExpression, attr: string, line = new LineInfo()) =>
        proxy(new Expression('attr', from, `"${attr}"`, line), Expression.proxyHandler);

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}
