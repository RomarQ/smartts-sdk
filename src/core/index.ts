import type { IToString, IToType } from '../typings';
import type { ILiteral } from '../typings/literal';
import type { IType } from '../typings/type';

import { Expression } from './expression';
import { Mutez } from './literal';
import Utils, { LineInfo } from '../misc/utils';
import { TUnit } from './type';

export class Flag {
    args: string[] = [];

    constructor(public flag: string, ...args: string[]) {
        this.args = args;
    }

    toString() {
        return `(${this.flag}${(' ' + this.args.join(' ')).trim()})`;
    }
}

interface EntryPointOptions {
    mock?: boolean;
    lazy?: boolean;
    lazyAndCodeless?: boolean;
}
export class EntryPoint {
    options = {
        mock: false,
        lazy: false,
        lazyAndCodeless: false,
    };
    inType?: IType;
    statements: IToString[] = [];

    constructor(public name: string, public line = new LineInfo()) {}

    config(options?: EntryPointOptions) {
        if (options?.mock) {
            this.options.mock = options.mock;
        }
        if (options?.lazy) {
            this.options.lazy = options.lazy;
        }
        if (options?.lazyAndCodeless) {
            this.options.lazyAndCodeless = options.lazyAndCodeless;
        }
        return this;
    }

    inputType(type: IType) {
        this.inType = type;
        return this;
    }

    code(callback: (arg: IToString) => IToString[]) {
        this.statements = callback(new Expression('params', new LineInfo()));
        return this;
    }

    toString() {
        const notMock = Utils.capitalizeBoolean(!this.options.mock);
        const isLazy = Utils.capitalizeBoolean(this.options.lazy);
        const isLazyAndCodeless = Utils.capitalizeBoolean(this.options.lazyAndCodeless);
        const hasParams = Utils.capitalizeBoolean(!!this.inType);
        return `(${this.name} ${notMock} ${isLazy} ${isLazyAndCodeless} ${hasParams} ${
            this.line
        } (${this.statements.join(' ')}))`;
    }
}

export class Contract {
    options: {
        initialBalance: ILiteral;
        flags: Flag[];
    } = {
        initialBalance: Mutez(0),
        flags: [],
    };
    public initialStorage: IToString & IToType;
    public entries: EntryPoint[];

    constructor(
        args: {
            initialStorage: IToString & IToType;
            entries: EntryPoint[];
        },
        public line = new LineInfo(),
    ) {
        this.initialStorage = args.initialStorage;
        this.entries = args.entries;
    }

    config(options?: { initialBalance?: ILiteral; flags?: Flag[] }) {
        if (options?.flags) {
            this.options.flags = options.flags;
        }
        if (options?.initialBalance) {
            this.options.initialBalance = options.initialBalance;
        }
        return this;
    }

    toString() {
        return `
        (
            template_id (static_id 0 ${this.line})
            storage ${this.initialStorage.toString()}
            storage_type (${this.initialStorage.toType()})
            messages (${this.entries.map((entry) => entry.toString()).join(' ')})
            flags (${this.options.flags.map((flag) => flag.toString())})
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance ${this.options.initialBalance}
        )
        `;
    }
}

export { default as Command } from './command';
export { default as Expression } from './expression';
export { default as Type } from './type';
export { default as Literal } from './literal';
export * from './blockchain_operations';
