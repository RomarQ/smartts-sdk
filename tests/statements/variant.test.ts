import { ForEachOf, MatchVariant, SetValue } from '../../src/statement';
import { ContractStorage, String } from '../../src/expression';
import { Contract, EntryPoint } from '../../src/core';
import { TList, TString, TVariant } from '../../src/type';
import { verifyContractCompilationOutput } from '../util';

describe('Test Variant statements', () => {
    it('MatchVariant 1', () => {
        const contract = new Contract().setStorage(String('')).addEntrypoint(
            new EntryPoint('ep1').inputType(TVariant({ action1: TString(), action2: TString() })).code((arg) => [
                MatchVariant(arg)
                    .Case('action1', (arg) => [SetValue(ContractStorage(), arg)])
                    .Case('action2', (arg) => [SetValue(ContractStorage(), arg)]),
            ]),
        );

        verifyContractCompilationOutput(contract);
    });
    it('MatchVariant 2', () => {
        const contract = new Contract().setStorage(String('')).addEntrypoint(
            new EntryPoint('ep1').inputType(TList(TVariant({ action1: TString(), action2: TString() }))).code((arg) => [
                ForEachOf(arg).Do((iter) => [
                    MatchVariant(iter)
                        .Case('action1', (payload) => [SetValue(ContractStorage(), payload)])
                        .Case('action2', (payload) => [SetValue(ContractStorage(), payload)]),
                ]),
            ]),
        );

        verifyContractCompilationOutput(contract);
    });
});
