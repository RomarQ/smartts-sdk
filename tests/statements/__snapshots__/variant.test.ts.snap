// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Variant statements MatchVariant 1 1`] = `
"
        (
            template_id (static_id 0 (\\"variant.test.ts\\" 9))
            storage (literal (string \\"\\") (\\"variant.test.ts\\" 10))
            storage_type (())
            messages ((ep1 True False False True (\\"variant.test.ts\\" 12) ((set_type (params (\\"variant.test.ts\\" 12)) (variant ((action1 \\"string\\") (action2 \\"string\\")) (Some ((\\"action1\\") (\\"action2\\"))) (\\"\\" -1)) (\\"variant.test.ts\\" 18)) (match_cases (params (\\"variant.test.ts\\" 12)) \\"__MATCH_1__\\" ((match (cases_arg \\"__MATCH_1__\\" (\\"variant.test.ts\\" 12)) action1 \\"__MATCH_1___action1\\" ((set (data) (variant_arg \\"__MATCH_1___action1\\" (\\"variant.test.ts\\" 12)) (\\"variant.test.ts\\" 12))) (\\"variant.test.ts\\" 12)) (match (cases_arg \\"__MATCH_1__\\" (\\"variant.test.ts\\" 12)) action2 \\"__MATCH_1___action2\\" ((set (data) (variant_arg \\"__MATCH_1___action2\\" (\\"variant.test.ts\\" 12)) (\\"variant.test.ts\\" 12))) (\\"variant.test.ts\\" 12))) (\\"variant.test.ts\\" 12)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"variant.test.ts\\" 9))
        )
        "
`;

exports[`Test Variant statements MatchVariant 1 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"string\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"or\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"string\\",
                        \\"annots\\": [
                            \\"%action1\\"
                        ]
                    },
                    {
                        \\"prim\\": \\"string\\",
                        \\"annots\\": [
                            \\"%action2\\"
                        ]
                    }
                ],
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CAR\\"
                },
                {
                    \\"prim\\": \\"IF_LEFT\\",
                    \\"args\\": [
                        [],
                        []
                    ]
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Variant statements MatchVariant 2 1`] = `
"
        (
            template_id (static_id 0 (\\"variant.test.ts\\" 24))
            storage (literal (string \\"\\") (\\"variant.test.ts\\" 25))
            storage_type (())
            messages ((ep1 True False False True (\\"variant.test.ts\\" 27) ((set_type (params (\\"variant.test.ts\\" 27)) (list (variant ((action1 \\"string\\") (action2 \\"string\\")) (Some ((\\"action1\\") (\\"action2\\"))) (\\"\\" -1))) (\\"variant.test.ts\\" 37)) (forGroup \\"__ITERATOR_FOREACH_2__\\" (params (\\"variant.test.ts\\" 27)) ((match_cases (iter \\"__ITERATOR_FOREACH_2__\\" (\\"variant.test.ts\\" 29)) \\"__MATCH_3__\\" ((match (cases_arg \\"__MATCH_3__\\" (\\"variant.test.ts\\" 29)) action1 \\"__MATCH_3___action1\\" ((set (data) (variant_arg \\"__MATCH_3___action1\\" (\\"variant.test.ts\\" 29)) (\\"variant.test.ts\\" 29))) (\\"variant.test.ts\\" 29)) (match (cases_arg \\"__MATCH_3__\\" (\\"variant.test.ts\\" 29)) action2 \\"__MATCH_3___action2\\" ((set (data) (variant_arg \\"__MATCH_3___action2\\" (\\"variant.test.ts\\" 29)) (\\"variant.test.ts\\" 29))) (\\"variant.test.ts\\" 29))) (\\"variant.test.ts\\" 29))) (\\"variant.test.ts\\" 29)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"variant.test.ts\\" 24))
        )
        "
`;

exports[`Test Variant statements MatchVariant 2 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"string\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"list\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"or\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"string\\",
                                \\"annots\\": [
                                    \\"%action1\\"
                                ]
                            },
                            {
                                \\"prim\\": \\"string\\",
                                \\"annots\\": [
                                    \\"%action2\\"
                                ]
                            }
                        ]
                    }
                ],
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"UNPAIR\\"
                },
                {
                    \\"prim\\": \\"DUP\\"
                },
                {
                    \\"prim\\": \\"ITER\\",
                    \\"args\\": [
                        [
                            {
                                \\"prim\\": \\"IF_LEFT\\",
                                \\"args\\": [
                                    [
                                        {
                                            \\"prim\\": \\"DIG\\",
                                            \\"args\\": [
                                                {
                                                    \\"int\\": \\"2\\"
                                                }
                                            ]
                                        },
                                        {
                                            \\"prim\\": \\"DROP\\"
                                        },
                                        {
                                            \\"prim\\": \\"SWAP\\"
                                        }
                                    ],
                                    [
                                        {
                                            \\"prim\\": \\"DIG\\",
                                            \\"args\\": [
                                                {
                                                    \\"int\\": \\"2\\"
                                                }
                                            ]
                                        },
                                        {
                                            \\"prim\\": \\"DROP\\"
                                        },
                                        {
                                            \\"prim\\": \\"SWAP\\"
                                        }
                                    ]
                                ]
                            }
                        ]
                    ]
                },
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;
